Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -0,0 +1,1011 @@
+      MolOps::AdjustDegree::HeavyDegree : MolOps::AdjustDegree::NoAdjust;
+  adjustQueryProperties(core, &adjustParams);
+  
+  for(std::map<int,int>::iterator it=labels.begin(); it!=labels.end(); ++it)
+    core.getAtomWithIdx(it->first)->setProp(RLABEL, it->second);
+
+  return true;
+}
+
+namespace {
+// RGroup Class to hold the attached bits
+struct RGroupData {
+  boost::shared_ptr<RWMol> combinedMol;
+  std::vector<boost::shared_ptr<ROMol> > mols;  // All the mols in the rgroup
+  std::set<std::string> smilesSet;              // used for rgroup equivalence
+  std::string smiles;  // smiles for all the mols in the rgroup (with attachments)
+  std::set<int> attachments;                    // attachment points
+  
+  void add(boost::shared_ptr<ROMol> newMol,
+           const std::vector<int> &rlabel_attachments) {
+    // some fragments can be add multiple times if they are cyclic
+    for (size_t i=0; i<mols.size(); ++i) {
+      if (newMol.get() == mols[i].get())
+        return;
+    }
+
+    std::copy( rlabel_attachments.begin(), rlabel_attachments.end(),
+               std::inserter( attachments, attachments.end() ) );
+
+    mols.push_back(newMol);
+    std::string smi = MolToSmiles(*newMol, true);
+    smilesSet.insert(smi);
+    if (!combinedMol.get()) {
+      combinedMol = boost::shared_ptr<RWMol>(new RWMol(*mols[0].get()));
+    } else {
+      ROMol *m = combineMols(*combinedMol.get(), *newMol.get());
+      combinedMol.reset(new RWMol(*m));
+      delete m;
+    }
+    smiles = getSmiles();
+  }
+
+  std::map<int,int> getNumBondsToRlabels() const {
+    std::map<int,int> rlabelsUsedCount;
+
+    for (ROMol::AtomIterator atIt = combinedMol->beginAtoms();
+         atIt != combinedMol->endAtoms();
+         ++atIt) {
+      Atom *atom = *atIt;
+      int rlabel; 
+      if(atom->getPropIfPresent<int>(RLABEL, rlabel))
+        rlabelsUsedCount[rlabel] += 1;
+    }
+    return rlabelsUsedCount;
+  }
+  
+  bool isHydrogen() const {  // is the rgroup all Hs
+    for(size_t i=0; i<mols.size(); ++i ) {
+      for (ROMol::AtomIterator atIt = mols[i]->beginAtoms();
+           atIt != mols[i]->endAtoms(); ++atIt) {
+        if ((*atIt)->getAtomicNum() > 1) return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  std::string getSmiles()
+      const {  // compute the canonical smiles for the attachments
+    std::string s;
+    for (std::set<std::string>::const_iterator it = smilesSet.begin();
+         it != smilesSet.end(); ++it) {
+      s += *it;
+    }
+    return s;
+  }
+};
+}
+
+namespace {
+struct RGroupMatch {
+  size_t core_idx;
+  std::map<int, RGroupData> rgroups;
+  RGroupMatch( size_t core_idx, const std::map<int, RGroupData> &rgroups ) :
+      core_idx(core_idx), rgroups(rgroups) {
+  }
+
+};
+
+struct CartesianProduct {
+  std::vector<size_t> permutation;
+  std::vector<size_t> sizes;
+  size_t maxPermutations;
+  size_t permutationCount;
+  CartesianProduct(const std::vector<size_t> &sizes)
```

Comment:
I agree, I got used to this hack for constructors that

```
class T {
   int foo;
  T(int foo) : foo(foo) {}
};
```

Actually works correctly and is unambiguous in the language.  However, it isn't in the brain so I should probably excise it.